[scroll-animations] scroll timelines should use the document's scrolling element as their default and root source

graouts opened the pull request:
#### 2d9d00fde2da68e148583639633bba2e871a73fc
<pre>
[scroll-animations] scroll timelines should use the document&apos;s scrolling element as their default and root source
<a href="https://bugs.webkit.org/show_bug.cgi?id=283412">https://bugs.webkit.org/show_bug.cgi?id=283412</a>
<a href="https://rdar.apple.com/140268956">rdar://140268956</a>

Reviewed by Simon Fraser.

Our scroll timeline code assumed that the document element was the root-most scroller in all cases,
whereas in quirks mode it is the &lt;body&gt; element. This manifested itself in the WPT test
`scroll-animations/scroll-timelines/animation-with-root-scroller.html` which runs in quirks mode
and we&apos;d fail to return a valid `ScrollableArea` for `&lt;body&gt;` because `scrollableAreaForSourceRenderer()`
would only match against the document element. As a result, that test&apos;s animation would never resolve
its `ready` promise since the scroll timeline would never return a resolved current time.

We make several changes to address this.

First, we get rid of `sourceElementForProgressCalculation()` and instead resolve the correct source in
the `source()` getter accounting for the `Scroller` value set for anonymous timelines (those originated
by a `scroll()` value in CSS). This means that anonymous timelines will also report the right source
through the JS API and not just within our internal source resolution.

Then in `scrollableAreaForSourceRenderer()` we correctly check for the document&apos;s `scrollingElement()`
instead of the document element to look for the right `ScrollableArea`.

While working in `scrollableAreaForSourceRenderer()`, we also move it to the protected section since
it&apos;s only used from within `ScrollTimeline` and its subclass `ViewTimeline`. Additionally, we make
its element parameter `const` and change the `Ref&lt;&gt;` parameter to a bare reference per the guidelines,
creating a `Ref` as needed within the function&apos;s body.

Finally, we extend the WPT test coverage to check the anonymous timeline source is correct in quirks mode
and non-quirks mode, including when the `&lt;body&gt;` can be scrolled independently from the document viewport.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/scroll-timeline-anonymous-source-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/scroll-timeline-anonymous-source-quirks-mode-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/scroll-timeline-anonymous-source-quirks-mode.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/scroll-timeline-anonymous-source-scrollable-body-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/scroll-timeline-anonymous-source-scrollable-body-quirks-mode-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/scroll-timeline-anonymous-source-scrollable-body-quirks-mode.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/scroll-timeline-anonymous-source-scrollable-body.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/scroll-timeline-anonymous-source.html: Added.
* Source/WebCore/animation/ScrollTimeline.cpp:
(WebCore::ScrollTimeline::source const):
(WebCore::ScrollTimeline::scrollableAreaForSourceRenderer):
(WebCore::ScrollTimeline::computeTimelineData const):
(WebCore::ScrollTimeline::sourceElementForProgressCalculation const): Deleted.
* Source/WebCore/animation/ScrollTimeline.h:
(WebCore::ScrollTimeline::source const): Deleted.

Canonical link: <a href="https://commits.webkit.org/286957@main">https://commits.webkit.org/286957@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/57cd791d5c01222f981a2455ca01d53ce6044888

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/77729 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/56762 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/30654 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/82373 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/29003 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/79846 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/65919 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/5063 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/60899 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/29003 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/80795 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/50866 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/66694 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/41190 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/48231 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/24205 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/27339 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/69347 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/24553 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/83735 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/5108 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/3448 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/83735 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/121/builds/5264 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/66681 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/83735 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/126/builds/12384 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/10474 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12030 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/5056 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/7834 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/5065 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/8503 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/6833 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/animation/ScrollTimeline.cpp:
```diff
@@ -194,13 +195,25 @@ void ScrollTimeline::setTimelineScopeElement(const Element& element)
         return nullptr;
 
-    if (renderer->element() == document->documentElement())
+    auto isRootScroller = [&] {
```
The test is using `scrollingElement` (and runs in quirks mode because it has no DOCTYPE) so this code also needs to test `scrollingElement`.

Also add a non-quirks mode test maybe?

smfr replied:
There should also be a test where the body is independently scrollable from the root.

smfr commented on Source/WebCore/animation/ScrollTimeline.cpp:
```diff
@@ -102,6 +102,34 @@ ScrollTimeline::ScrollTimeline(Scroller scroller, ScrollAxis axis)
+    switch (m_scroller) {
+    case Scroller::Nearest: {
+        if (CheckedPtr subjectRenderer = m_source->renderer()) {
+            if (RefPtr source = subjectRenderer->enclosingScrollableContainer()->element()) {
```
You should null-check the result of `enclosingScrollableContainer()`.

graouts replied:
We haven't seen a crash yet, so maybe we should `ASSERT()`, but I'll add null check if you don't see a reason why we can add an `ASSERT()` here.

smfr commented on Source/WebCore/animation/ScrollTimeline.cpp:
```diff
@@ -102,6 +102,34 @@ ScrollTimeline::ScrollTimeline(Scroller scroller, ScrollAxis axis)
+                // make sure to return Document::scrollingElement() in case the document element is
+                // returned by enclosingScrollableContainer() but it was not explicitly set as the source.
+                RefPtr documentElement = document->documentElement();
+                if (m_source.get() != documentElement && source == documentElement)
```
Confusing to have `source` and `m_source`.

smfr replied:
I still find this code a bit confusing. Can you have an early return for the non-root case, and deal with the documentElement/scrollingElement case in its own code block?

graouts replied:
> Confusing to have `source` and `m_source`.

Indeed! I'll rename `source` to `nearestSource` since this is what we're looking for here.

graouts replied:
> I still find this code a bit confusing. Can you have an early return for the non-root case, and deal with the documentElement/scrollingElement case in its own code block?

Will try in an updated version of this patch, let me know if you have further comments.

graouts committed:
[scroll-animations] scroll timelines should use the document's scrolling element as their default and root source
https://bugs.webkit.org/show_bug.cgi?id=283412
rdar://140268956

Reviewed by Simon Fraser.

Our scroll timeline code assumed that the document element was the root-most scroller in all cases,
whereas in quirks mode it is the <body> element. This manifested itself in the WPT test
`scroll-animations/scroll-timelines/animation-with-root-scroller.html` which runs in quirks mode
and we'd fail to return a valid `ScrollableArea` for `<body>` because `scrollableAreaForSourceRenderer()`
would only match against the document element. As a result, that test's animation would never resolve
its `ready` promise since the scroll timeline would never return a resolved current time.

We make several changes to address this.

First, we get rid of `sourceElementForProgressCalculation()` and instead resolve the correct source in
the `source()` getter accounting for the `Scroller` value set for anonymous timelines (those originated
by a `scroll()` value in CSS). This means that anonymous timelines will also report the right source
through the JS API and not just within our internal source resolution.

Then in `scrollableAreaForSourceRenderer()` we correctly check for the document's `scrollingElement()`
instead of the document element to look for the right `ScrollableArea`.

While working in `scrollableAreaForSourceRenderer()`, we also move it to the protected section since
it's only used from within `ScrollTimeline` and its subclass `ViewTimeline`. Additionally, we make
its element parameter `const` and change the `Ref<>` parameter to a bare reference per the guidelines,
creating a `Ref` as needed within the function's body.

Finally, we extend the WPT test coverage to check the anonymous timeline source is correct in quirks mode
and non-quirks mode, including when the `<body>` can be scrolled independently from the document viewport.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/scroll-timeline-anonymous-source-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/scroll-timeline-anonymous-source-quirks-mode-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/scroll-timeline-anonymous-source-quirks-mode.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/scroll-timeline-anonymous-source-scrollable-body-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/scroll-timeline-anonymous-source-scrollable-body-quirks-mode-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/scroll-timeline-anonymous-source-scrollable-body-quirks-mode.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/scroll-timeline-anonymous-source-scrollable-body.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/scroll-timeline-anonymous-source.html: Added.
* Source/WebCore/animation/ScrollTimeline.cpp:
(WebCore::ScrollTimeline::source const):
(WebCore::ScrollTimeline::scrollableAreaForSourceRenderer):
(WebCore::ScrollTimeline::computeTimelineData const):
(WebCore::ScrollTimeline::sourceElementForProgressCalculation const): Deleted.
* Source/WebCore/animation/ScrollTimeline.h:
(WebCore::ScrollTimeline::source const): Deleted.

Canonical link: https://commits.webkit.org/286957@main
